File: frontend/src/lib/calculo/selecao_celula.ts
/**
 * Selecao da celula recomendada para o dimensionamento.
 * Port fiel de backend/app/calculo/selecao_celula.py.
 */

import type { OpcaoCelulaCalc } from "./series_paralelo";
import type { ResumoDimensionamento } from "./dimensionamento";

/**
 * Seleciona a celula recomendada dentre as opcoes calculadas.
 *
 * Modo "Automatica":
 *   - Filtra opcoes com cont_pack >= i_max E capacidade_pack >= ah_necessario.
 *   - Ordena por: menor paralelo → menor capacidade_pack → menor peso_pack.
 *   - Retorna a primeira; null se catalogo vazio ou sem opcao valida.
 *
 * Outro valor:
 *   - Selecao manual: retorna a opcao cujo id seja "{fabricante} {ah}Ah".
 *   - Retorna null se nao encontrado (em vez de propagar excecao).
 */
export function escolherCelula(
  modo: string,
  opcoes: OpcaoCelulaCalc[],
  resumo: ResumoDimensionamento,
): OpcaoCelulaCalc | null {
  if (modo === "Automática") {
    const validas = opcoes.filter(
      (o) =>
        o.cont_pack >= resumo.i_max &&
        o.capacidade_pack >= resumo.ah_necessario,
    );
    if (!validas.length) return null;
    return [...validas].sort(
      (a, b) =>
        a.paralelo - b.paralelo ||
        a.capacidade_pack - b.capacidade_pack ||
        a.peso_pack - b.peso_pack,
    )[0];
  }

  return opcoes.find((o) => `${o.fabricante} ${o.ah}Ah` === modo) ?? null;
}

File: backend/app/calculo/selecao_celula.py
"""
Selecao da celula recomendada para o dimensionamento.

Equivale a funcao `escolher_celula` do app Streamlit atual.
"""

from typing import List, Mapping, Optional


def escolher_celula(modo: str, opcoes: List[dict], resumo: Mapping) -> Optional[dict]:
    """
    Seleciona a celula recomendada entre as opcoes calculadas para o
    catalogo completo (uma opcao por celula, ja com serie/paralelo
    calculados por `series_paralelo.calcular_opcao_celula`).

    Regras:
    - Modo "Automatica" (valor literal "Automática", igual ao app atual):
      filtra as opcoes cuja corrente continua do pack (FullEnergy, ja com a
      margem MARGEM_CONTINUA) seja >= corrente maxima exigida (i_max) E cuja
      capacidade do pack seja >= ah_necessario (capacidade_pack >=
      ah_necessario, sem tolerancia). Entre as validas, ordena por:
        1. menor paralelo;
        2. menor capacidade_pack;
        3. menor peso_pack;
      e retorna a primeira. Se nao houver opcao valida (ex.: catalogo
      vazio), retorna None.
    - Qualquer outro valor de `modo` e tratado como selecao manual: retorna a
      opcao cujo identificador "{fabricante} {ah}Ah" seja igual a `modo`. Se
      nao houver correspondencia, propaga StopIteration -- mesmo
      comportamento do app atual (nenhum tratamento adicional de erro).
    """
    if modo == "Automática":
        ah_necessario = resumo["ah_necessario"]

        validas = [
            o for o in opcoes
            if o["cont_pack"] >= resumo["i_max"]
            and o["capacidade_pack"] >= ah_necessario
        ]
        if not validas:
            return None
        return sorted(
            validas,
            key=lambda x: (
                x["paralelo"],
                x["capacidade_pack"],
                x["peso_pack"],
            ),
        )[0]

    return next(o for o in opcoes if modo == f"{o['fabricante']} {o['ah']}Ah")
